File: webservices/src/main/java/fieg/modulos/interfacecobranca/enums/IntegraProtheus.java
package fieg.modulos.interfacecobranca.enums;

import com.fasterxml.jackson.databind.JsonNode;
import fieg.core.exceptions.ValorInvalidoException;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public enum IntegraProtheus {

    CONTRATO_PROTHEUS('C', "Contrato Protheus"),
    INTEGRACAO_FINANCEIRA('F', "Integração Financeira"),
    ISENTO('I', "Título isento"),
    VENDA_PARCELADA('P', "Venda Parcelada"),
    VENDA_DIRETA('V', "Venda Direta"),
    VENDA_AVULSA_PARCELADA('A', "Venda Avulsa Parcelada"),
    CONTRATO_PROTHEUS_25('T', "Contrato Protheus 25");

    private final Character letra;
    private final String descricao;

    IntegraProtheus(Character letra, String descricao) {
        this.letra = letra;
        this.descricao = descricao;
    }

    public Character getLetra() {
        return letra;
    }

    public String getDescricao() {
        return descricao;
    }

    public static Optional<IntegraProtheus> getByLetra(Character letra) {
        return Arrays.stream(values()).filter(it -> it.getLetra().equals(letra)).findFirst();
    }

    public static List<IntegraProtheus> integracoesFinanceira() {
        return List.of(IntegraProtheus.INTEGRACAO_FINANCEIRA, IntegraProtheus.CONTRATO_PROTHEUS_25);
    }

    public static List<IntegraProtheus> convertJsonNodeToList(JsonNode node) throws ValorInvalidoException {
        List<IntegraProtheus> lista = new ArrayList<>();

        if (!node.isArray()) {
            throw new ValorInvalidoException("Valores Integra Protheus deve ser um array");
        }

        for (JsonNode element : node) {
            String nome = element.asText();
            IntegraProtheus integraProtheus = IntegraProtheus.valueOf(nome);
            lista.add(integraProtheus);
        }

        return lista;
    }

}

File: web/src/models/DTOs/Protheus/IntegraProtheus.ts
export interface IntegraProtheusDTO {
  letra: string;
  descricao: string;
  nomeEnum: string;
}

interface IntegraProtheusType {
  CONTRATO_PROTHEUS: IntegraProtheusDTO;
  INTEGRACAO_FINANCEIRA: IntegraProtheusDTO;
  ISENTO: IntegraProtheusDTO;
  VENDA_PARCELADA: IntegraProtheusDTO;
  VENDA_DIRETA: IntegraProtheusDTO;
  VENDA_AVULSA_PARCELADA: IntegraProtheusDTO;
  CONTRATO_PROTHEUS_25: IntegraProtheusDTO;
}

export type IntegraProtheus = keyof IntegraProtheusType;

export const integraProtheusValues: IntegraProtheusType = {
  CONTRATO_PROTHEUS: {
    letra: "C",
    descricao: "Contrato Protheus",
    nomeEnum: "CONTRATO_PROTHEUS",
  },
  INTEGRACAO_FINANCEIRA: {
    letra: "F",
    descricao: "Integração Financeira",
    nomeEnum: "INTEGRACAO_FINANCEIRA",
  },
  ISENTO: {
    letra: "I",
    descricao: "Título isento",
    nomeEnum: "ISENTO",
  },
  VENDA_PARCELADA: {
    letra: "P",
    descricao: "Venda Parcelada",
    nomeEnum: "VENDA_PARCELADA",
  },
  VENDA_DIRETA: {
    letra: "V",
    descricao: "Venda Direta",
    nomeEnum: "VENDA_DIRETA",
  },
  VENDA_AVULSA_PARCELADA: {
    letra: "A",
    descricao: "Venda Avulsa Parcelada",
    nomeEnum: "VENDA_AVULSA_PARCELADA",
  },
  CONTRATO_PROTHEUS_25: {
    letra: "T",
    descricao: "Contrato Protheus 25",
    nomeEnum: "CONTRATO_PROTHEUS_25",
  },
};
